safeopt/gp_opt.py

```diff
@@ -567,3 +582,352 @@ def get_maximum(self, context=None):
+    """
+    A class to maximize a function using SafeOpt with a Particle Swarm Optimization
+    heuristic.
+    Note that it doesn't support the use of a Lipschitz constant
+
+    Parameters
+    ----------
+    gp: GPy Gaussian process
+        A Gaussian process which is initialized with safe, initial data points.
+        If a list of GPs then the first one is the value, while all the
+        other ones are safety constraints.
+    fmin: list of floats
+        Safety threshold for the function value. If multiple safety constraints
+        are used this can also be a list of floats (the first one is always
+        the one for the values, can be set to None if not wanted)
+    beta: float or callable
+        A constant or a function of the time step that scales the confidence
+        interval of the acquisition function.
+    threshold: float
+        The algorithm will not try to expand any points that are below this
+        threshold. This makes the algorithm stop expanding points eventually.
+    scaling: list of floats or "auto"
+        A list used to scale the GP uncertainties to compensate for
+        different input sizes. This should be set to the maximal variance of each kernel.
+        You should probably leave this to "auto" unless your kernel is non stationnary
+    bounds: pair of floats or list of pairs of floats
+        If a list is given, then each pair represents the lower/upper bound in
+        each dimension.
+        Otherwise, we assume the same bounds for all dimensions
+    swarm_size: int
+        The number of particles in each of the optimization swarms
+    verbose: boolean
+        If set to True, the algorithm will print some information about the optimization process
+
+    """
+
+    def __init__(self, gp, fmin, beta=3.0, num_contexts=0, threshold=0,
+                 scaling='auto', bounds=(-5, 5), swarm_size=20, verbose=False):
+        super(SafeOptSwarmSoft, self).__init__(function, gp, safe_set, beta)
+
+        self.fmin = fmin
+        if not isinstance(self.fmin, list):
+            self.fmin = [self.fmin] * len(self.gps)
+        self.fmin = np.atleast_1d(np.asarray(self.fmin).squeeze())
+
+        # Safe set
+        self.S = self.gps[0].X
+
+        self.swarm_size = swarm_size
+        self.max_iters = 100  # number of swarm iterations
+
+        self.verbose = verbose
+
+        if not isinstance(self.fmin, list):
+            self.bounds = [bounds] * self.S.shape[1]
+        else:
+            self.bounds = bounds
+
+        self.var_max = gp.kern.K(np.atleast_2d(
+            gp.X[-1, :]), np.atleast_2d(gp.X[-1, :]))
+
+        # These are estimates of the best lower bound, and its location
+        self.best_lower_bound = -np.inf
+        self.greedy_point = self.S[0, :]
+
+    # this function compute the value of each particles, depending on the
+    # swarm type
+    def compute_particle_fitness(self, particles, swarm_type):
+        beta = self.beta(self.t)
```

Comment:
You might want to make particle positions part of the class (e.g., `self.particles`)